### DISCUSSION TITLE
How to delete a file from a range of commits if it was created in an earlier commit?

### DISCUSSION THREAD

**Discussion by xlexle** [newren/git-filter-repo#447] [Q&A]
Hello,

It took me a while to realize that my attempts to delete a file from a specified range of `--refs` was not working because the file was created earlier. Example command:

```
git-filter-repo --invert-paths --path FILE --force --refs HEAD~5..
```

This deletes FILE only if FILE was created in the last 5 commits. If the file was created in HEAD~6, git-filter-repo happily processes all of the last 5 commits without any notification that FILE was not deleted. I don't know if this should be a bug report or if this is intended behavior.

My question is, how do I go about deleting FILE from the range HEAD\~5..HEAD if FILE was created in HEAD\~6 or earlier?

**Comment by newren** [comment]
You'd have to use a --commit-callback, and add an extra FileChange for the deletion of the file in question.  It'd look something like this:

git filter-repo --refs HEAD~5.. --commit-callback "commit.file_changes += [
    FileChange(b'D', b'PATH/TO/FILENAME')]"

Technically, my command is trying to delete PATH/TO/FILENAME from every commit in the range `HEAD~5..HEAD`, which might feel slightly like overkill (if you have a linear history, you only need it to be deleted in `HEAD~4`), but deleting a file that didn't exist in the parent commit is translated by fast-import to a no-op, so it doesn't hurt anything.
